smithy-model/src/main/java/software/amazon/smithy/model/selector/WrappedSelector.java

```diff
@@ -47,23 +51,93 @@
         }
     }
 
-    @Override
-    public Runner runner() {
-        return new Runner(delegate, startingShapeType);
-    }
-
     @Override
     public String toString() {
         return expression;
     }
 
     @Override
     public boolean equals(Object other) {
-        return other instanceof WrappedSelector && expression.equals(((WrappedSelector) other).expression);
+        return other instanceof Selector && toString().equals(other.toString());
     }
 
     @Override
     public int hashCode() {
         return expression.hashCode();
     }
+
+    @Override
+    public Set<Shape> select(Model model) {
+        Set<Shape> result = new HashSet<>();
+        // This is more optimized that using shapes() and collecting to a Set
+        // because it avoids creating streams and buffering the result of
+        // pushing each shape into internal selectors.
+        pushShapes(model, (ctx, s) -> {
+            result.add(s);
+            return true;
+        });
+        return result;
+    }
+
+    @Override
+    public void consumeMatches(Model model, Consumer<ShapeMatch> shapeMatchConsumer) {
+        // This is more optimized that using matches() and collecting to a Set
+        // because it avoids creating streams and buffering the result of
+        // pushing each shape into internal selectors.
+        pushShapes(model, (ctx, s) -> {
+            // Note: ctx.copyVars is not used because ShapeMatch automatically
+            // creates a copy of the variables in its constructor.
+            shapeMatchConsumer.accept(new ShapeMatch(s, ctx.getVars()));
+            return true;
+        });
+    }
+
+    @Override
+    public Stream<Shape> shapes(Model model) {
+        Context context = createContext(model);
+        return streamStartingShape(model).flatMap(shape -> {
+            List<Shape> result = new ArrayList<>();
+            delegate.push(context.clearVars(), shape, (ctx, s) -> {
+                result.add(s);
+                return true;
+            });
+            return result.stream();
+        });
+    }
+
+    @Override
+    public Stream<ShapeMatch> matches(Model model) {
+        Context context = createContext(model);
+        return streamStartingShape(model).flatMap(shape -> {
+            List<ShapeMatch> result = new ArrayList<>();
+            delegate.push(context.clearVars(), shape, (ctx, s) -> {
+                result.add(new ShapeMatch(s, ctx.copyVars()));
```

Comment:
Is there a reason this uses `copyVars` with the above note on using `getVars` instead in `consumeMatches`? I don't see anything different enough to warrant it.

Reply:
Oh good point. Will change this one.